### ISSUE TITLE
Optional route parameters are not yet supported

### ISSUE THREAD

**Issue by maximus-lynn** [vercel/next.js#22342]
**What version of Next.js are you using?**

latest

**What version of Node.js are you using?**

14

**What browser are you using?**

n/a

**What operating system are you using?**

macOS

**How are you deploying your application?**

n/a

**Describe the Bug**

I've been trying to convert our CRA to Next.js but it is a large project and need to do it slowly rather than in one go. So I read your documentation on how to migrate and come across the below url:

https://nextjs.org/docs/migrating/from-create-react-app#single-page-app-spa

I thought this is great, exactly what I need. New pages can be put into next.js whilst old ones staying in CRA until refactored. 

I've tried to implement this but I'm running into the error below:
```
Optional route parameters are not yet supported ("[[…app]]").
```

So got three questions here. 
1. Is this not supported in your latest version? 
2. if it isn't should it be removed from your documentation?
3. Have I missed a trick and doing something wrong?

Many thanks

**Expected Behavior**

For Optional route params to be supported.

**To Reproduce**

Try to use [[...app]] in a project

**Comment by LuudJanssen** [comment]
Hi @maximus-lynn, I created an new project with `create-next-app` and changed the default `index.js` to `[[...app]].js` and everything works fine, so it must be something specific to your project which is not working.

Are you sure your filename is something like `[[...app]].js` including the triple dots (`...`)? If not, you'll get the error you showed:

```
Error: Optional route parameters are not yet supported ("[[app]]").
```

As to why this specific case is not supported, I don't know, but you can always create an `index.js` file which catches the "/" route file and re-exports the render function in `[app].js`.

Does this answer your question?

**Comment by maximus-lynn** [comment]
hmm, very possible I've missed something.

Don't suppose you could push up your test somewhere to see where I might of gone wrong?

**Comment by LuudJanssen** [comment]
@maximus-lynn, [here ya go](https://codesandbox.io/s/nextjs-optional-route-parameters-vu5o5?file=/pages/%5B%5B...app%5D%5D.js). Any route you visit will be catched by the `[[...app]].js` page. If you want to test it, go to **any** route (e.g. `/test/test/test`) in the sandbox and the same thing will render.

**Comment by maximus-lynn** [comment]
Sorry to be a pain but I think it'll help other people. 

Do you have an example where you use this method to wrap all of the react-router routes in CRA?

**Comment by LuudJanssen** [comment]
Hi @maximus-lynn. No problem! I don't have an example as I have little experience with CRA and react-router and it would take too much time to create the example.

If you follow [the Migrating from Create React App tutorial](https://nextjs.org/docs/migrating/from-create-react-app) and [the Migrating from React Router tutorial](https://nextjs.org/docs/migrating/from-react-router), what problems are you running into? Maybe you can share an example repository and show what isn't working for you?

Also, you might consider moving this to [a discussion](https://github.com/vercel/next.js/discussions), as this probably isn't related to a bug, but a support request, and you might get some more helpful answers there.

**Comment by maximus-lynn** [comment]
Yeah thanks, 100% a discussion now :)

**Comment by Aeolun** [comment]
I have exactly the same issue. Follow the documentation exactly and end up with an unhelpful error message. That by itself should be a bug, since catch all route parameters have been supported for a while, at the very least the message should be something like 'invalid route parameters'.

Here's a repro: https://github.com/Aeolun/brokentest

But honestly, when I committed that it became abundantly clear what the issue is: `new file:   "pages/[[\342\200\246app]].tsx"`

The documentation should probably be updated to not use (or automatically convert to) fancy triple dots.

**Comment by simonespa** [comment]
@LuudJanssen @maximus-lynn @Aeolun I managed to reproduce this error.

Optional routes parameters only works when you use an "Optional catch-all" route (e.g. [[...slug]].tsx) https://nextjs.org/docs/routing/dynamic-routes#optional-catch-all-routeshttps://nextjs.org/docs/routing/dynamic-routes#optional-catch-all-routes.

It doesn't work otherwise, if you use a simple dynamic route such as `/path/to/[something]/[[optional]].tsx` which matches `/path/to/*/*` with no arbitrary nesting. You could use the catch-all such as `/path/to/[[...somethingOptional]].tsx` which matches `/path/to/*`, `/path/to/*/*`, `/path/to/*/*/*`, ... , etc. but it is not the same, given every sub path is also captured. One thing you could do is to programmatically stop the nesting at the second level and programmatically send a 404 for everything else.

To sum up, if the error thrown on the console by Next.js is ` Optional route parameters are not yet supported`, I think this issue can be transformed in a feature request, unless it is already in the roadmap and captured somewhere else.

**Comment by Aeolun** [comment]
@simonespa I’m sorry, maybe my previous comment wasn’t clear.

The issue for me was that I copied the ‘catch-all’ example from the documentation, but the documentation uses a unicode ‘triple dot’ character instead of 3 actual dots.

That is one issue (with the documentation), but the real issue is just that the error message doesn’t make any sense. It should say something like: *Invalid route.*

**Comment by yordis** [comment]
> The issue for me was that I copied the ‘catch-all’ example from the documentation, but the documentation uses a unicode ‘triple dot’ character instead of 3 actual dots, and that makes the route invalid.

@Aeolun thank you so much, your comment saved me a lot of time

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
